Fix/440 Per-image customization for quality, size, and style in Image Generation

faisal-alvi opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

This PR enhances the Image Generation feature by allowing users to modify image quality, size, and style directly from the media modal. Previously, these options were only configurable at the global settings level, requiring users to change settings manually before generating an image and then reverting them if needed.  

![classifai-440-01](https://github.com/user-attachments/assets/b472b5ad-8be7-425d-ad0d-143ee47a1ef4)

- Added controls in the media modal for adjusting image quality (standard/HD), size (1024x1024, 1792x1024, 1024x1792), and style (vivid/natural) per image generation.  
- Defaults to global settings, ensuring consistency while allowing flexibility for individual use cases.  
- Introduced a global setting to disable per-image adjustments.

![image](https://github.com/user-attachments/assets/a77a27b8-416b-45b2-b3e4-973dc1c2acf2)

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #440

### How to test the Change

1. Navigate to ClassifAI Image Generation Settings and note the global settings for **quality, size, and style**.
2. Navigate to the image generation section. Ensure that the quality, size, and style options default to the global settings.  
3. In the media modal, change the **quality (standard/HD)**, **size (1024x1024, 1792x1024, 1024x1792)**, and **style (vivid/natural)**. Generate an image and verify that it reflects the selected options.  
4. In ClassifAI Image Generation Settings, toggle the new setting to **disable per-image adjustments**. Open the media modal again and verify that the optios for quality, size, and style are now hidden.
5. If per-image settings are disabled, generate an image and confirm that it uses the **global settings** instead.  
6. Also ensure that the same behaviour occurs in both image generation pages, 1) Under Media and 2) While adding an image in edit post page.

### Changelog Entry

> Added - Per-image customization for quality, size, and style in Image Generation.

### Credits

Props @dkotter @jeffpaul @faisal-alvi 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added [Critical Flows, Test Cases, and/or End-to-End Tests](https://10up.github.io/Open-Source-Best-Practices/testing/) to cover my change.
- [ ] All new and existing tests pass.

faisal-alvi committed:
add `get_image_quality_options()`

faisal-alvi committed:
add `get_image_size_options()`

faisal-alvi committed:
add `get_image_style_options()`

faisal-alvi committed:
add HTML markup

faisal-alvi committed:
show settings on checkbox click

faisal-alvi committed:
pass params in `makeRequest`

faisal-alvi committed:
transfer the additional data

faisal-alvi committed:
detect the settings and pass

faisal-alvi committed:
add a global settings to native settings page

faisal-alvi committed:
add a global settings to block-based settings page

faisal-alvi committed:
show additional options as per the global settings

faisal-alvi committed:
prefix with "Enable"

faisal-alvi committed:
tweaks for edit post view compat

faisal-alvi committed:
remove js code

faisal-alvi committed:
CSS + handle toggle in CSS

faisal-alvi committed:
eslint fixes,
/home/runner/work/classifai/classifai/src/js/features/image-generation/media-modal/collections/images.js
Error:   16:2  error  Missing JSDoc @param "quality" declaration  jsdoc/require-param
Error:   16:2  error  Missing JSDoc @param "size" declaration     jsdoc/require-param
Error:   16:2  error  Missing JSDoc @param "style" declaration    jsdoc/require-param

/home/runner/work/classifai/classifai/src/js/features/image-generation/media-modal/views/generated-images-container.js
Error:   31:31  error  Replace `·this.prompt,·this.quality,·this.size,·this.style·` with `⏎↹↹↹this.prompt,⏎↹↹↹this.quality,⏎↹↹↹this.size,⏎↹↹↹this.style⏎↹↹`  prettier/prettier

✖ 4 problems (4 errors, 0 warnings)
  4 errors and 0 warnings potentially fixable with the `--fix` option.

Error: Process completed with exit code 1.

faisal-alvi committed:
phpcs fixes,
FILE: includes/Classifai/Features/ImageGeneration.php
-----------------------------------------------------------------------------------------------
FOUND 6 ERRORS AND 0 WARNINGS AFFECTING 6 LINES
-----------------------------------------------------------------------------------------------
 316 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
 318 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
 328 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
 330 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
 340 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
 342 | ERROR | Line indented incorrectly; expected 8 tabs, found 9 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
-----------------------------------------------------------------------------------------------
Error: Process completed with exit code 1.

faisal-alvi committed:
eslint error for vertical inline,
/home/runner/work/classifai/classifai/src/js/features/image-generation/media-modal/collections/images.js
Error:   16:2  error  Expected JSDoc block lines to be aligned  jsdoc/check-line-alignment

dkotter commented on includes/Classifai/Features/ImageGeneration.php:
```diff
@@ -271,10 +271,11 @@ public function print_media_templates() {
+		$settings           = $this->get_settings();
+		$provider_id        = $settings['provider'];
+		$number_of_images   = absint( $settings[ $provider_id ]['number_of_images'] );
+		$per_image_settings = isset( $settings[ $provider_id ]['per_image_settings'] ) ? $settings[ $provider_id ]['per_image_settings'] : false;
```
```suggestion
		$per_image_settings = $settings[ $provider_id ]['per_image_settings'] ?? false;
```

dkotter commented on includes/Classifai/Features/ImageGeneration.php:
```diff
@@ -299,6 +300,52 @@ public function print_media_templates() {
+					<br>
+					<input type="checkbox" id="view-additional-image-generation-settings" />
+					<label id="view-additional-image-generation-settings-label" for="view-additional-image-generation-settings">
+						<?php esc_html_e( 'View Additional Settings', 'classifai' ); ?>
```
```suggestion
						<?php esc_html_e( 'Additional settings', 'classifai' ); ?>
```

dkotter commented on src/js/settings/components/provider-settings/openai-dalle.js:
```diff
@@ -157,6 +158,21 @@ export const OpenAIDallESettings = ( { isConfigured = false } ) => {
 				/>
 			</SettingsRow>
+			<SettingsRow
+				label={ __( 'Enable per-image settings', 'classifai' ) }
```
```suggestion
				label={ __( 'Per-image settings', 'classifai' ) }
```

dkotter commented on src/js/settings/components/provider-settings/openai-dalle.js:
```diff
@@ -157,6 +158,21 @@ export const OpenAIDallESettings = ( { isConfigured = false } ) => {
+			<SettingsRow
+				label={ __( 'Enable per-image settings', 'classifai' ) }
+				description={ __(
+					'Enable this to allow users to select the quality, size, and style of the generated image.',
```
```suggestion
					'If enabled, allows users to select the quality, size, and style when generating an image.',
```

dkotter commented on includes/Classifai/Providers/OpenAI/DallE.php:
```diff
@@ -149,15 +142,65 @@ public function render_provider_fields() {
 			]
 		);
+
+		add_settings_field(
```
With settings now managed in JS, we don't need to use `add_settings_field` anymore. We left that in place for existing settings but for new settings, I suggest we don't add this as it isn't actually used

dkotter replied:
The one thing we do need though that I'm not seeing in this PR is adding this new setting to our defaults, in `get_default_provider_settings`

dkotter commented on src/js/features/image-generation/media-modal/views/prompt.js:
```diff
@@ -35,17 +35,31 @@ const Prompt = wp.media.View.extend( {
 			prompt = event.target.parentElement
 				.querySelector( '.prompt' )
 				.value.trim();
+			quality = event.target.parentElement
+				.querySelector( '.quality' )
+				.value.trim();
+
+			size = event.target.parentElement
+				.querySelector( '.size' )
+				.value.trim();
+
+			style = event.target.parentElement
+				.querySelector( '.style' )
+				.value.trim();
```
Because we are now using `event.target.parentElement` multiple times, we should store that in a variable instead of getting that element over and over

dkotter reviewed (changes requested):
Overall code looks good here (haven't tested it yet) but left a few comments that need addressing.

In addition, I'm seeing E2E tests failing so those need looked at

faisal-alvi committed:
suggestions implemented

faisal-alvi committed:
adding default for pre-image setting

faisal-alvi committed:
eslint fixes

faisal-alvi committed:
try to fix e2e; making generic selector

faisal-alvi committed:
eslint fix,
/home/runner/work/classifai/classifai/src/js/features/image-generation/media-modal/views/prompt.js
Error:   46:46  error  Delete `⏎↹↹↹↹`  prettier/prettier
Error:   49:48  error  Delete `⏎↹↹↹↹`  prettier/prettier
Error:   52:42  error  Delete `⏎↹↹↹↹`  prettier/prettier
Error:   55:44  error  Delete `⏎↹↹↹↹`  prettier/prettier

/home/runner/work/classifai/classifai/tests/cypress/integration/image-processing/image-generation-openai-dalle.test.js
  0:0  warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override

✖ 5 problems (4 errors, 1 warning)
  4 errors and 0 warnings potentially fixable with the `--fix` option.

faisal-alvi committed:
revert test change

faisal-alvi committed:
fix E2E

faisal-alvi committed:
keep dropdowns to pass default values, fixing e2e

faisal-alvi committed:
fix button position
